### PR TITLE
Add annotation to place pci addresses on root bus

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

If a VMI is started with the annotation
"kubevirt.io/placePCIDevicesOnRootComplex": "true", all PCI devices are
attached to the root PCI bus. This makes them not hot-unplugable in
windows. Otherwise normal users get big popups on Windows 10 which present them disks, interfaces and all other kinds of devices for unplug. The unplug operation as non-admin user also succeeds and leaves the VMs in broken states.

In order to achive the placement of all PCI addresses on the root bus, the following things need to be done:

 * Place all our devices on pci addresses like `<address type='pci' slot='N' domain='0' bus='0' function='0')/>`
 * Do the same for implicit default devices which get implicitly added if we add new guest-visible devices (this makes it a little bit brittle)

The functional test verifies that no `pcie-root-port` element is there. If we see such an element we know that a default device which we overlooked was added by libvirt to this non-root bus. If we place all devices upfront on `domain='0' bus='0' function='0' slot='N'`, libvirt will not add extra pci controllers.

Regarding to available slots on the root bus, the following restrictions apply:

* the bus 0 has 32 slots total. 
* slot 0 is reserved
* slot 0x1f has a hardcoded (by qemu) SATA controller
* slot 1 is usually used for the primary video device
* slot 0x1b is used for the first ich9 sound card

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by vladikr** [comment]
This probably shouldn't work in a combination with a custom PCI devices that devices (virtio disks and interfaces) can provide via `pciAddress` field.

Also, what about SRIOV? There are coming with their own addresses.

**Comment by rmohr** [comment]
> This probably shouldn't work in a combination with a custom PCI devices that devices (virtio disks and interfaces) can provide via `pciAddress` field.
> 

Yes, I am unsure about these. I guess the code should make an exception for specific address requests.

> Also, what about SRIOV? There are coming with their own addresses.

I mistakenly changed the PCI address on the host dev. In the new version it just gets a different PCI address than on the host, but it should still work as expected.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 1]
```diff
@@ -457,6 +457,9 @@ func Convert_v1_Input_To_api_InputDevice(input *v1.Input, inputDevice *Input, _
 	}
 
 	inputDevice.Bus = input.Bus
+	if inputDevice.Bus == "" {
+		inputDevice.Bus = "virtio"
+	}
```
Comment (vladikr) [inline pkg/virt-launcher/virtwrap/api/converter.go]: there is 
```
if input.Bus != "virtio" && input.Bus != "usb" {
	input.Bus = "usb"
}
```
just before this, so I guess it'll always be `usb` if it wasn't set to `virtio` here

Comment (rmohr) [reply pkg/virt-launcher/virtwrap/api/converter.go]: yes that is interesting ...

Comment (rmohr) [reply pkg/virt-launcher/virtwrap/api/converter.go]: oh man, I added that code. I thought it was there before. Thanks @vladikr  :+1:

Comment (rmohr) [reply pkg/virt-launcher/virtwrap/api/converter.go]: fixed.

**Comment by stu-gott** [comment]
Looks like there's an issue in the unit tests

**Comment by rmohr** [comment]
> Looks like there's an issue in the unit tests

thx, yes the ppc64 domain xml does not match ...

**Comment by rmohr** [comment]
/hold

a ppc64 unit test needs to be adjusted.

**Comment by rmohr** [comment]
/hold cancel

fixed the unit tests and rebased. @stu-gott could you have another look?

**Comment by rmohr** [comment]
@murilofv FYI the third PR in this series partially addresses #3067. Here we have a good example on why we should always run all unit tests for all archs at the same time. WIthout this I would always have had to push to travis to see if I finally fixed the test xmls. If you like would be nice to get a review from you too.

**Comment by murilofv** [comment]
> @murilofv FYI the third PR in this series partially addresses #3067. Here we have a good example on why we should always run all unit tests for all archs at the same time. WIthout this I would always have had to push to travis to see if I finally fixed the test xmls. If you like would be nice to get a review from you too.

Oh cool, thank you for the work on it! I am not familiar with Ginkgo at all so I wasn't aware of the mechanism to do it that way. From my layman's POV it looks good though.

File: pkg/virt-launcher/virtwrap/manager.go [hunk 2]
```diff
@@ -1380,9 +1380,18 @@ func (l *LibvirtDomainManager) ListAllDomains() ([]*api.Domain, error) {
...
 }
 
-func (l *LibvirtDomainManager) setDomainSpecWithHooks(vmi *v1.VirtualMachineInstance, spec *api.DomainSpec) (cli.VirDomain, error) {
+func (l *LibvirtDomainManager) setDomainSpecWithHooks(vmi *v1.VirtualMachineInstance, origSpec *api.DomainSpec) (cli.VirDomain, error) {
 
+	spec := origSpec.DeepCopy()
 	hooksManager := hooks.GetManager()
+
+	if val := vmi.Annotations[v1.PlacePCIDevicesOnRootComplex]; val == "true" {
+		err := api.PlacePCIDevicesOnRootComplex(spec)
```
Comment (davidvossel) [inline pkg/virt-launcher/virtwrap/manager.go]: why isn't this done in the `Convert_v1_VirtualMachine_To_api_Domain()` function instead?

This technically works the way you have it, however I'd be surprised to find out the domain is getting mutated during the "set" function like this.

File: pkg/virt-launcher/virtwrap/api/converter.go [hunk 3]
```diff
@@ -926,6 +926,12 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
 			Index: "0",
 			Model: "none",
 		})
+	} else {
+		domain.Spec.Devices.Controllers = append(domain.Spec.Devices.Controllers, Controller{
+			Type:  "usb",
+			Index: "0",
+			Model: "qemu-xhci",
```
Comment (davidvossel) [inline pkg/virt-launcher/virtwrap/api/converter.go]: should we only be defining the usb controller if a usb device exists?

Comment (rmohr) [reply pkg/virt-launcher/virtwrap/api/converter.go]: yes, that is what we do. Note the `Model: "none"` above.

File: pkg/virt-launcher/virtwrap/manager.go (hunk 2 above)
Comment (rmohr) [reply pkg/virt-launcher/virtwrap/manager.go]: Addressed.

**Comment by rmohr** [comment]
/hold  cancel

merge conflict resolved

**Comment by rmohr** [comment]
would be grate if someone can put a lgtm again.

**Comment by rmohr** [comment]
/override pull-kubevirt-e2e-k8s-1.16.2

that job does not exist anymore.

**Comment by rmohr** [comment]
/override pull-kubevirt-e2e-kind-k8s-sriov-1.17.0

that job does  not work at the moment.

**Comment by rmohr** [comment]
/override pull-kubevirt-e2e-k8s-1.15-ceph
/override pull-kubevirt-e2e-k8s-1.16

jobs are stuck.

**Pull request merged**